aggregator/generic_elem.go

```diff
@@ -0,0 +1,315 @@
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package aggregator
+
+import (
+	"sync"
+	"time"
+
+	raggregation "github.com/m3db/m3aggregator/aggregation"
+	maggregation "github.com/m3db/m3metrics/aggregation"
+	"github.com/m3db/m3metrics/metric/id"
+	"github.com/m3db/m3metrics/metric/unaggregated"
+	"github.com/m3db/m3metrics/op/applied"
+	"github.com/m3db/m3metrics/policy"
+
+	"github.com/mauricelam/genny/generic"
+)
+
+type lockedAggregation interface {
+	generic.Type
+	sync.Locker
+
+	// Add adds a new metric value.
+	Add(mu unaggregated.MetricUnion)
+
+	// ValueOf returns the value for the given aggregation type.
+	ValueOf(aggType maggregation.Type) float64
+
+	// Close closes the aggregation object.
+	Close()
+}
+
+type genericElemPool interface {
+	generic.Type
+
+	// Put returns an element to a pool.
+	Put(value *GenericElem)
+}
+
+type genericElemBase interface {
+	generic.Type
+
+	// FullPrefix returns the full prefix for the given metric type.
+	FullPrefix(opts Options) []byte
+
+	// DefaultAggregationTypeStrings returns the default aggregation type strings.
+	DefaultAggregationTypeStrings(aggTypesOpts maggregation.TypesOptions) [][]byte
+
+	// DefaultAggregationTypes returns the default aggregation types.
+	DefaultAggregationTypes(aggTypesOpts maggregation.TypesOptions) maggregation.Types
+
+	// TypeStringFor returns the type string for the given aggregation type.
+	TypeStringFor(aggTypesOpts maggregation.TypesOptions, aggType maggregation.Type) []byte
+
+	// ElemPool returns the pool for the given element.
+	ElemPool(opts Options) genericElemPool
+
+	// NewLockedAggregation creates a new locked aggregation.
+	NewLockedAggregation(opts Options, aggOpts raggregation.Options) lockedAggregation
+
+	// ResetSetData resets and sets data.
+	ResetSetData(
+		aggTypesOpts maggregation.TypesOptions,
+		aggTypes maggregation.Types,
+		useDefaultAggregation bool,
+	)
+
+	// Close closes the element.
+	Close()
+}
+
+type timedAggregation struct {
+	timeNanos   int64
+	aggregation lockedAggregation
+}
+
+func (ta *timedAggregation) Reset() {
+	ta.timeNanos = 0
+	ta.aggregation = nil
+}
+
+// GenericElem is an element storing time-bucketed aggregations.
+type GenericElem struct {
+	elemBase
+	genericElemBase
```

Comment:
both `elemBase` and `genericElemBase` sound very generic and they sounded similar, maybe rename `genericElemBase` to `genericTypedElemBase`?

Reply:
Hm I initially called it `metricTypeElemBase` but landed on just `genericElemBase`. I can rename to `typeSpecificElemBase` sure.